client/core/types.go

```diff
@@ -735,6 +737,25 @@ func (a *dexAccount) ID() account.AccountID {
 	return a.id
 }
 
+func (a *dexAccount) hasKeys() bool {
+	a.keyMtx.RLock()
+	defer a.keyMtx.RUnlock()
+	return len(a.encKey) > 0
+}
+
+// isRegistered is true if fee is paid (legacy) or initial bond is posted, even
+// if the payment/bond is yet to be confirmed.
+func (a *dexAccount) isRegistered() bool {
+	a.authMtx.RLock()
+	defer a.authMtx.RUnlock()
+
+	return a.isAuthed || // 'connect' succeeded, so fee is paid or initial bond is posted+confirmed
+		len(a.feeCoin) > 0 || // fee is paid but either yet to 'connect' or is waiting confirmations
+		len(a.bonds) > 0 || // initial bond is posted+confirmed, probably hasn't 'connect'ed yet
+		len(a.pendingBonds) > 0 || // initial bond is posted, waiting confirmations
+		len(a.expiredBonds) > 0 // initial bond _was_ posted and confirmed
```

Comment:
Refunded bonds appear in none of these slices, so this `isRegistered` can have false negatives.  I'm toying with a `registered` flag for `dexAccount` that is set according to DB info.  See this `AccountInfo` method and toggling to true when register or bondconfirmed succeeds: https://github.com/decred/dcrdex/commit/91d612432c553c786d5d3647d358ec12beace507

Reply:
I initially thought that was a commit from one of your bond in-progress work but later realized it was a commit on top of my last commit in this PR. Tried adapting it and ensuring there are no cracks in the logic when I realized that the purpose of the `acct.isRegistered()` method is to prevent duplicate legacy registration and possibly to also prevent needless account discovery.

The only 3 callers of that method use the returned value to determine if to
- display the "create account to trade" button on the frontend
- prevent account discovery attempts
- prevent legacy registration attempts

For each of these cases, the `isRegistered` value would need to also consider pending payments (fee / bond txs that have been broadcast but not yet confirmed).

There's also a potential for false negatives with the db bonds check, for a registered account that was recovered with no active bonds, but that's less of a concern. The db `AccountInfo.Registered()` method would return false if the account has no legacy fee payment and no active bonds reported by the server on recovery, until and unless the user posts another bond.